docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,320 @@
+  extender multiple times (e.g. because of other constraints not being met)
+
+### The runtime
+
+Once we know where to run the container, it's time to set up its environment. At
+a minimum, we'll need to map the host device(s) into the container. Because each
+manufacturer exposes different device nodes (`/dev/ati/card0`, `/dev/nvidia0`,
+but also the required `/dev/nvidiactl` and `/dev/nvidia-uvm`), some of the logic
+needs to be hardware-specific, mapping from a logical device to a list of device
+nodes necessary for software to talk to it.
+
+Support binaries and libraries are often versioned along with the kernel module,
+so there should be further hooks to project those under `/bin` and some kind of
+`/lib` before the application is started. This can be done for Docker with the
+use of a versioned [Docker
+volume](https://docs.docker.com/engine/userguide/containers/dockervolumes/). The
+alternative would be for each application to bundle everything in their image.
+
+#### NVIDIA support
+
+The first implementation and proof of concept will be for NVIDIA devices, the
+setup that is most common and with more extensive infrastructure, namely the
+[nvidia-docker plugin](https://github.com/NVIDIA/nvidia-docker). The plugin is
+an officially supported solution, thus avoiding a lot of new low level code, as
+it takes care of functionality such as:
+
+- creating a Docker volume with binaries such as `nvidia-smi` and shared
+libraries
+- providing HTTP endpoints that monitoring tools can use to collect GPU metrics
+- abstracting details such as `/dev` entry names for each device, as well as
+control ones like `nvidiactl`
+
+The `nvidia-docker` wrapper also verifies that the CUDA version required by a
+given image is supported by the host drivers, through inspection of well-known
+image labels, if present. We should try to provide equivalent checks, either
+for CUDA or OpenCL.
+
+This is current sample output from `nvidia-docker-plugin`, wrapped for
+readability:
+
+    $ curl -s localhost:3476/docker/cli
+    --device=/dev/nvidiactl --device=/dev/nvidia-uvm --device=/dev/nvidia0
+    --volume-driver=nvidia-docker
+    --volume=nvidia_driver_352.68:/usr/local/nvidia:ro
+
+It runs as a daemon listening for HTTP requests on port 3476. The endpoint above
+returns flags that need to be added to the Docker command line in order to
+expose GPUs to the containers. There are optional URL arguments to request
+specific devices if more than one are present on the system, as well as specific
+versions of the support software.
+
+An obvious improvement is an additional endpoint for JSON output. The unresolved
+question is whether `nvidia-docker-plugin` should run standalone as it does
+today (called over HTTP, perhaps with endpoints for a new Kubernetes resource
+API) or whether the relevant code from its `nvidia` package should be linked
+directly into kubelet. A partial list of tradeoffs:
```

Comment:
I see. The unit of resource sharing is one device, am I right? More specifically, CUDA device.

Previously I was wondering how runtime could isolate devices and cores for each container. IIUC, kubelet will be responsible for managing device sharing. It looks like docker nvidia doesn't support such mapping yet -- that plugin is just to make the nvidia driver agnostic. I think that's why we need to query the NVIDIA for device information. As a side note, this is quite different than how we are dealing with CPU: (not very sure) we are using container runtime API to create containers and let container runtime handle isolation. Underlying is cgroup doing isolation.